### DISCUSSION TITLE
Where can cfg.data.test be used to test a model?

### DISCUSSION THREAD

**Discussion by julianzille** [open-mmlab/mmpose#1663] [Q&A]
Hi everyone. I am busy using mmpose for an animal pose-estimation project (largely following AP10K config/annotation) format.

I have a dataset split into train, val and test in my config file. I can track where the train and val datasets are used with 
`train_model(model, datasets, cfg, distributed=False, validate=True, meta=dict())`
But I cannot see if/where the test dataset (cfg.data.test) is used anywhere... am I correct that its intended use is by functions in mmpose/apis/test.py? If so, I have been trying to test the model with cfg.data.test using single_gpu_test and multi_gpu_test as follows:
```
from mmpose.datasets import build_dataset, build_dataloader
from mmpose.apis import single_gpu_test, init_pose_model, multi_gpu_test
from torch.utils.data import DataLoader

latest_model = init_pose_model(cfg,'work_dirs/hrnet_w32_acino_256x256_epoch50/latest.pth')

test_dataset = build_dataset(cfg.data.test)

test_dataloader = build_dataloader(test_dataset,32,4)

test_results=multi_gpu_test(latest_model,test_dataloader)
```

This yields an assertion error: 
```
...
...
File /notebooks/mmpose/models/detectors/top_down.py:171, in TopDown.forward_test(self, img, img_metas, return_heatmap, **kwargs)
    166         """Defines the computation performed at every call when testing."""
    167 #         print()
    168 #         print(img.size(0))
    169 #         print(len(img_metas))
--> 171         assert img.size(0) == len(img_metas)
    172         batch_size, _, img_height, img_width = img.shape
    173         if batch_size > 1:

AssertionError: 
```
I am not entirely sure what 'img' and 'img_metas' are or how they work, but img.size(0) == 32 and len(img_metas) == 1. 
Am I making an obvious error? Is this the correct approach?

**Comment by ly015** [comment]
Thanks for using MMPose. Only `data.train` and `data.val` are used during training, and `data.test` is used during testing. Please refer to `tools/test.py` for details.
